### ISSUE TITLE
Home Energy Meter - Gen5 (ZW095), sensor ignored in case of production exceeding consumption

### ISSUE THREAD

**Issue by pawcio50501** [domoticz/domoticz#6180]
### Background
Problem detected after upgrade from domoticz 2022 to 2024 
My property is generating more energy than consumed.

### MQTT data

zwave2domoticz/sensor/House_total/electric_kwh_value_3/config
{"value_template":"{{ value_json.value }}","unit_of_measurement":"kWh","state_class":"total_increasing","device_class":"energy","state_topic":"zwave/18/50/3/value/65537","availability":[{"payload_available":"true","payload_not_available":"false","topic":"zwave/18/status","value_template":"{{'true' if value_json.value else 'false'}}"},{"topic":"zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/status","value_template":"{{'online' if value_json.value else 'offline'}}"},{"payload_available":"true","payload_not_available":"false","topic":"zwave/driver/status"}],"availability_mode":"all","json_attributes_topic":"zwave/18/50/3/value/65537","device":{"identifiers":["zwavejs2mqtt_0xc9cacfd6_node18"],"manufacturer":"AEON Labs","model":"Home Energy Meter - Gen5 (ZW095)","name":"House total","sw_version":"1.4"},"name":"House total_electric_kwh_value_3","unique_id":"zwavejs2mqtt_0xc9cacfd6_18-50-3-value-65537"}

zwave/18/50/3/value/65537
{"time":1728406392283,"value":-6487.507,"nodeName":"House total","nodeLocation":"Garage"}

### Source of multiple problems:
MQTTAutoDiscover::GuessSensorTypeValue

**Comment by gizmocuz** [comment]
Hi, thank you for reporting an issue.
Could you elaborate more on what the problem is? What does not go correctly? And maybe you have a fix as well?

**Comment by pawcio50501** [comment]
Hi,
I can see at least two problems:
1. https://github.com/domoticz/domoticz/blob/d13355bcc34df7f1d6624be4dc69644497be4d3f/hardware/MQTTAutoDiscover.cpp#L1874 
For some reason values below -1MWh are recognized as an invalid one, which is obviously not proper. In my case it is now -6MWh
2. https://github.com/domoticz/domoticz/blob/d13355bcc34df7f1d6624be4dc69644497be4d3f/hardware/MQTTAutoDiscover.cpp#L1914
Newly received values are lower than the previous, which leads in updating epoch value, which in turn leads in wrong dkWh calculation.

**Comment by waltervl** [comment]
So both lines are related to https://github.com/domoticz/domoticz/pull/6163
 @dwmw2 any thoughts?

**Comment by dwmw2** [comment]
Hm, this sensor describes itself as `total_increasing`, which is described here:
https://developers.home-assistant.io/blog/2021/08/16/state_class_total/

If the value reported by the sensor goes backwards, we're *supposed* to assume it reset to zero and add the current reading to the total. That's what it means when it tells us that it's a `total_increasing` sensor.

So this looks like a bug in zwave2domoticz. But we do need to work around it. Perhaps just disregard the `total_increasing` for sensors where the device ID contains `zwavejs2mqtt`?

**Comment by dwmw2** [comment]
(The -1MWh limit isn't new in #6163; all I did there was change the name of the variable and fix some spelling)

**Comment by pawcio50501** [comment]
ZW095 can be configured in the following ways:
- absolute value of energy
- positive/negative power -algebraic sum of energy (my selection)
- positive power only
- negative power only

**Comment by dwmw2** [comment]
Unless the output is truly monotonic, it shouldn't be describing itself as `total_increasing`. That's for things like power meters which count up, but might sometimes get reset to zero, and we expect Domoticz to keep counting from where it was before. Like an old-style pulse counter that just counted the 1Wh LED pulses on the front of a non-smart meter.

**Comment by pawcio50501** [comment]
Note: I've also configured HA. Both HA and domoticz are using same mqtt broker, both receive same data, but only HA is able to display proper values.

**Comment by dwmw2** [comment]
Huh, that implies that HA isn't following its *own* documentation for how such devices should behave...

**Comment by dwmw2** [comment]
Can you work out if HA *ever* does what its documentation describes, or if it also has some kind of opt-out for devices which are known to advertise `total_increasing` when they shouldn't, or other heuristics?

**Comment by gizmocuz** [comment]
Is this the only config object? Or are there multiple? For example, are there two more with the total power pushed back and used?

For instance, total power used clamp_1, total power delivered clamp_1, total power used clamp_2, total power delivered clamp_2

If not, how would you know how much you have used or delivered?

Do you have the clamp on the output? Or on the cable that is delivering (like solar panels), and the grid usage?

If you put the clamp on the main output, yes that can go up/down, nice to see the actual usage (plus or minus), but this way you cannot measure the total up/down

**Comment by pawcio50501** [comment]
[HA-mqtt-270de645f0fbd631d27daccb5c515b9d-House total-96520a2dd93f4839e984a3a2310fa083.json](https://github.com/user-attachments/files/17349111/HA-mqtt-270de645f0fbd631d27daccb5c515b9d-House.total-96520a2dd93f4839e984a3a2310fa083.json)
Please find attached diagnostics from the HA - you will see all configs and values received for this one specific device
PS. it's not the only measure device in my setup. I don't have any problems to see exact data regarding production and consumption on separete graphs.

**Comment by gizmocuz** [comment]
I browsed through this log file, and it seems you have multiple kWh sensors... 

House total House total_electric_kwh_value
House total House total_electric_kwh_value_1
House total House total_electric_kwh_value_2
House total House total_electric_kwh_value_3

Are the others working correctly?

**Comment by pawcio50501** [comment]
value_1 to 3 corresponds to specific phases (I have 3 phase installation)
value is just a sum of the 3 above.
Works only one which have "state" above -1MWh:
"entity_id": "sensor.house_total_house_total_electric_kwh_value_1",
            "state": "-382.86",

Rest are ignored:
"entity_id": "sensor.house_total_house_total_electric_kwh_value",
            "state": "-8282.101",
"entity_id": "sensor.house_total_house_total_electric_kwh_value_2",
            "state": "-1401.472", 
"entity_id": "sensor.house_total_house_total_electric_kwh_value_3",
            "state": "-6497.769",

**Comment by gizmocuz** [comment]
@pawcio50501 , is it possible you could try beta 16296?

**Comment by pawcio50501** [comment]
Hi, I did quick test updating MQTTAutoDiscover.* only to the latest version - still it is not working

**Comment by waltervl** [comment]
@pawcio50501 Still an issue?

**Comment by dwmw2** [comment]
Coming back to this because I'd really like to get #6163 merged again. I think this is a combination of two things.

Firstly, this code *always* rejected values of less than -1MWh so people with lots of export (yay!) would have suffered that already. I think we need to rethink that sanity check.

The other part I think is a zwavejs2mqtt bug, which would now potentially be in ZWave-JS UI? The issue was that it reported these devices as [`total_increasing`](https://developers.home-assistant.io/blog/2021/08/16/state_class_total/) when they were not. Meters which can handle export are *not* monotonically increasing. @robertslando should I file bugs?

**Comment by robertsLando** [comment]
> Meters which can handle export are not monotonically increasing

Yes and if possible also submitting a PR, the issue should be in this function: https://github.com/zwave-js/zwave-js-ui/blob/master/api/lib/Constants.ts#L120

**Comment by dwmw2** [comment]
> Yes and if possible also submitting a PR

Will attempt to get ZWave-JS-UI running again and do so, thanks.

For now I've made https://github.com/domoticz/domoticz/pull/6426 just not believe `total_increasing` for zwave devices.
